### ISSUE TITLE
Cloudflare is messing with reports

### ISSUE THREAD

**Issue by fcwill** [fcwill/fchat-pidgin#98]
I just wanted to report a conversation when I got the following error:

```
(14:48:43) flist: User filed a report against 'X': 'Posterity'
------------- LOG -------------
<Log redacted>

-----------------------------
(14:48:43) flist: cgi writing key, value: ticket, fct_X
(14:48:43) flist: cgi writing key, value: log, redacted

(14:48:43) flist: cgi writing key, value: channel, X
(14:48:43) flist: cgi writing key, value: reportText, Posterity
(14:48:43) flist: cgi writing key, value: account, X
(14:48:43) flist: cgi writing key, value: character, Nelwill
(14:48:43) flist: cgi writing key, value: reportUser, X
(14:48:43) flist: posting (len: 3017): <redacted>
(14:48:43) flist: HTTP-Request: POST /json/api/report-submit.php HTTP/1.1

Connection: close

User-Agent: Pidgin F-Chat 0.5.0

Accept: */*

Host: www.f-list.net

Content-Type: application/x-www-form-urlencoded

Content-Length: 3017



<redacted>
(14:48:43) util: requesting to fetch a URL
(14:48:43) util: Defaulting max download from https://www.f-list.net/json/api/report-submit.php to 524288
(14:48:43) dnsquery: Performing DNS lookup for www.f-list.net
(14:48:43) dnsquery: IP resolved for www.f-list.net
(14:48:43) proxy: Attempting connection to 104.24.25.27
(14:48:43) proxy: Connecting to www.f-list.net:443 with no proxy
(14:48:43) proxy: Connection in progress
(14:48:43) proxy: Connecting to www.f-list.net:443.
(14:48:43) proxy: Connected to www.f-list.net:443.
(14:48:43) nss: SSL version 3.3 using 128-bit AES-GCM with 128-bit AEAD MAC
Server Auth: 256-bit ECDSA, Key Exchange: 256-bit ECDHE, Compression: NULL
Cipher Suite Name: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
(14:48:43) nss: subject=CN=ssl255860.cloudflaressl.com,OU=PositiveSSL Multi-Domain,OU=Domain Control Validated issuer=CN=COMODO ECC Domain Validation Secure Server CA 2,O=COMODO CA Limited,L=Salford,ST=Greater Manchester,C=GB
(14:48:43) nss: partial certificate chain
(14:48:43) certificate/x509/tls_cached: Starting verify for www.f-list.net
(14:48:43) certificate/x509/tls_cached: Checking for cached cert...
(14:48:43) certificate/x509/tls_cached: ...Found cached cert
(14:48:43) nss/x509: Loading certificate from C:\Pidgin\.purple\certificates\x509\tls_peers\www.f-list.net
(14:48:43) certificate/x509/tls_cached: Peer cert matched cached
(14:48:43) nss/x509: Exporting certificate to C:\Pidgin\.purple\certificates\x509\tls_peers\www.f-list.net
(14:48:43) util: Writing file C:\Pidgin\.purple\certificates\x509\tls_peers\www.f-list.net
(14:48:43) nss: Trusting CN=ssl255860.cloudflaressl.com,OU=PositiveSSL Multi-Domain,OU=Domain Control Validated
(14:48:43) certificate: Successfully verified certificate for www.f-list.net
(14:48:43) util: request constructed
(14:48:43) util: Response headers: 'HTTP/1.1 403 Forbidden

Server: cloudflare-nginx

Date: Sun, 06 Mar 2016 13:48:44 GMT

Content-Type: text/html; charset=UTF-8

Transfer-Encoding: chunked

Connection: close

Set-Cookie: __cfduid=X; expires=Mon, 06-Mar-17 13:48:44 GMT; path=/; domain=.f-list.net; HttpOnly

Cache-Control: max-age=15

Expires: Sun, 06 Mar 2016 13:48:59 GMT

X-Frame-Options: SAMEORIGIN

CF-RAY: X



'
(14:48:43) flist: Web Request JSON Received: <!DOCTYPE html>
<!--[if lt IE 7]> <html class="no-js ie6 oldie" lang="en-US"> <![endif]-->
<!--[if IE 7]>    <html class="no-js ie7 oldie" lang="en-US"> <![endif]-->
<!--[if IE 8]>    <html class="no-js ie8 oldie" lang="en-US"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en-US"> <!--<![endif]-->
<head>
<title>HTTP hostname and TLS SNI hostname mismatch | CloudFlare</title></title>
<meta charset="UTF-8" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1" />
<meta name="robots" content="noindex, nofollow" />
<meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1" />
<link rel="stylesheet" id="cf_styles-css" href="/cdn-cgi/styles/cf.errors.css" type="text/css" media="screen,projection" />
<!--[if lt IE 9]><link rel="stylesheet" id='cf_styles-ie-css' href="/cdn-cgi/styles/cf.errors.ie.css" type="text/css" media="screen,projection" /><![endif]-->
<style type="text/css">body{margin:0;padding:0}</style>
<!--[if lte IE 9]><script type="text/javascript" src="/cdn-cgi/scripts/jquery.min.js"></script><![endif]-->
<!--[if gte IE 10]><!--><script type="text/javascript" src="/cdn-cgi/scripts/zepto.min.js"></script><!--<![endif]-->
<script type="text/javascript" src="/cdn-cgi/scripts/cf.common.js"></script>

</head>
<body>
  <div id="cf-wrapper">
    <div class="cf-alert cf-alert-error cf-cookie-error" id="cookie-alert" data-translate="enable_cookies">Please enable cookies.</div>
    <div id="cf-error-details" class="cf-error-details-wrapper">
      <div class="cf-wrapper cf-header cf-error-overview">
        <h1>
          <span class="cf-error-type" data-translate="error">Error</span>
          <span class="cf-error-code">1013</span>
          <small class="heading-ray-id">Ray ID: X &bull; 2016-03-06 13:48:44 UTC</small>
        </h1>
        <h2 class="cf-subheadline" data-translate="error_desc">HTTP hostname and TLS SNI hostname mismatch</h2>
      </div><!-- /.header -->

      <section></section><!-- spacer -->

      <div class="cf-section cf-wrapper">
        <div class="cf-columns two">
          <div class="cf-column">
            <h2 data-translate="what_happened">What happened?</h2>
            <p>You've requested an IP address that is part of the <a data-orig-proto="https" data-orig-ref="www.cloudflare.com/5xx-error-landing?utm_source=error_100x" target="_blank">CloudFlare</a> network. The request's Host header does not match the request's TLS SNI Host header.</p>
          </div>

          
          <div class="cf-column">
            <h2 data-translate="what_can_i_do">What can I do?</h2>
            <p>If you are interested in learning more about CloudFlare, please <a data-orig-proto="https" data-orig-ref="www.cloudflare.com/5xx-error-landing?utm_source=error_100x" target="_blank">visit our website</a>.</p>
          </div>
          
        </div>
      </div><!-- /.section -->

      <div class="cf-error-footer cf-wrapper">
  <p>
    <span class="cf-footer-item">CloudFlare Ray ID: <strong>X</strong></span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span data-translate="your_ip">Your IP</span>: X</span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span data-translate="performance_security_by">Performance &amp; security by</span> <a data-orig-proto="https" data-orig-ref="www.cloudflare.com/5xx-error-landing?utm_source=error_footer" id="brand_link" target="_blank">CloudFlare</a></span>
    
  </p>
</div><!-- /.error-footer -->


    </div><!-- /#cf-error-details -->
  </div><!-- /#cf-wrapper -->

  <script type="text/javascript">
  window._cf_translation = {};
  
  
</script>

</body>
</html>

(14:48:43) flist: Expected JSON Object, but did not parse with error message: <none>:1: Parse error: unexpected character `<', expected keyword
(14:48:43) flist: Raw JSON: <!DOCTYPE html>
<!--[if lt IE 7]> <html class="no-js ie6 oldie" lang="en-US"> <![endif]-->
<!--[if IE 7]>    <html class="no-js ie7 oldie" lang="en-US"> <![endif]-->
<!--[if IE 8]>    <html class="no-js ie8 oldie" lang="en-US"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en-US"> <!--<![endif]-->
<head>
<title>HTTP hostname and TLS SNI hostname mismatch | CloudFlare</title></title>
<meta charset="UTF-8" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1" />
<meta name="robots" content="noindex, nofollow" />
<meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1" />
<link rel="stylesheet" id="cf_styles-css" href="/cdn-cgi/styles/cf.errors.css" type="text/css" media="screen,projection" />
<!--[if lt IE 9]><link rel="stylesheet" id='cf_styles-ie-css' href="/cdn-cgi/styles/cf.errors.ie.css" type="text/css" media="screen,projection" /><![endif]-->
<style type="text/css">body{margin:0;padding:0}</style>
<!--[if lte IE 9]><script type="text/javascript" src="/cdn-cgi/scripts/jquery.min.js"></script><![endif]-->
<!--[if gte IE 10]><!--><script type="text/javascript" src="/cdn-cgi/scripts/zepto.min.js"></script><!--<![endif]-->
<script type="text/javascript" src="/cdn-cgi/scripts/cf.common.js"></script>

</head>
<body>
  <div id="cf-wrapper">
    <div class="cf-alert cf-alert-error cf-cookie-error" id="cookie-alert" data-translate="enable_cookies">Please enable cookies.</div>
    <div id="cf-error-details" class="cf-error-details-wrapper">
      <div class="cf-wrapper cf-header cf-error-overview">
        <h1>
          <span class="cf-error-type" data-translate="error">Error</span>
          <span class="cf-error-code">1013</span>
          <small class="heading-ray-id">Ray ID: X &bull; 2016-03-06 13:48:44 UTC</small>
        </h1>
        <h2 class="cf-subheadline" data-translate="error_desc">HTTP hostname and TLS SNI hostname mismatch</h2>
      </div><!-- /.header -->

      <section></section><!-- spacer -->

      <div class="cf-section cf-wrapper">
        <div class="cf-columns two">
          <div class="cf-column">
            <h2 data-translate="what_happened">What happened?</h2>
            <p>You've requested an IP address that is part of the <a data-orig-proto="https" data-orig-ref="www.cloudflare.com/5xx-error-landing?utm_source=error_100x" target="_blank">CloudFlare</a> network. The request's Host header does not match the request's TLS SNI Host header.</p>
          </div>

          
          <div class="cf-column">
            <h2 data-translate="what_can_i_do">What can I do?</h2>
            <p>If you are interested in learning more about CloudFlare, please <a data-orig-proto="https" data-orig-ref="www.cloudflare.com/5xx-error-landing?utm_source=error_100x" target="_blank">visit our website</a>.</p>
          </div>
          
        </div>
      </div><!-- /.section -->

      <div class="cf-error-footer cf-wrapper">
  <p>
    <span class="cf-footer-item">CloudFlare Ray ID: <strong>X</strong></span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span data-translate="your_ip">Your IP</span>: X</span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span data-translate="performance_security_by">Performance &amp; security by</span> <a data-orig-proto="https" data-orig-ref="www.cloudflare.com/5xx-error-landing?utm_source=error_footer" id="brand_link" target="_blank">CloudFlare</a></span>
    
  </p>
</div><!-- /.error-footer -->


    </div><!-- /#cf-error-details -->
  </div><!-- /#cf-wrapper -->

  <script type="text/javascript">
  window._cf_translation = {};
  
  
</script>

</body>
</html>

(14:48:43) flist: Log upload failed! Error Message: Invalid JSON.
```

This does not happen all the time, my first report worked just fine. Don't know how to reproduce, yet.
More information: https://support.cloudflare.com/hc/en-us/articles/204872628-Error-1013-HTTP-hostname-and-TLS-SNI-hostname-mismatch

**Comment by fcwill** [comment]
I think deleting all f-list certificates from .purple/certificates resolves this and other SSL issues.
I'll leave this open for a while for confirmation.

**Comment by syldrathecat** [comment]
@vobe

I assume this is the issue to talk about the 403 errors all over the place.

I did some tests and everything the f-chat plugin was doing looked completely normal. Nothing I changed with the HTTP requests seemed to make any difference.

Then I tried having pidgin load the ssl-gnutls plugin instead of the ssl-nss plugin and the errors suddenly stopped.

To make sure it wasn't anything weird with the plugin I made this stand-alone libpurple client to test with: https://gist.github.com/syldrathecat/7243fbf5163122b9bde9, and found the following:

* The problem is very consistently and only caused while using the NSS plugin,
* The problem never occurs with the GnuTLS plugin.
* The problem still occurs even if all requests are made sequentially.
* The problem does not occur if only one domain is ever requested from!

I took a closer look at the last point, noting that www.f-list.net and static.f-list.net share a common set of IP addresses and...

Adding the following lines to /etc/hosts caused the problem to disappear (note: don't copy these as they're likely CloudFlare IP addresses far from where you live):

    104.24.24.27    www.f-list.net
    104.24.25.27    static.f-list.net

And doing the opposite -- forcing them to have the same IP address -- causes the problem to happen 100% of the time, even with only 2 requests.

In Wireshark the connections are made individually with correct server_name fields. However, the second request is made with a Session ID (assuming this is to quickly re-establish SSL connections), which I guess is done incorrectly by NSS due to the IP addresses matching, but (again assuming) should NOT be done if the server_name field doesn't match.

I guess the question is: which plugins are various Pidgin distributions using by default? (Not sure about Windows. On Debian it's NSS).

Unfortunately plain HTTP requests to buddy icons redirect to HTTPS, so that's not a usable workaround unless they're willing to change that. I'm not sure of any other workarounds right now, short of manually looking up www.f-list.net and using the IP address to connect directly without SNI (Cloudflare will accept this from what I've seen, but it's very likely to break in the future).

Will keep digging and possibly end up filing a report upstream somewhere.

**Comment by syldrathecat** [comment]
I've taken a look deeper and RFC 6066 says this about SNI:

    When the server is deciding whether or not to accept a request to
    resume a session, the contents of a server_name extension MAY be used
    in the lookup of the session in the session cache.  The client SHOULD
    include the same server_name extension in the session resumption
    request as it did in the full handshake that established the session.
    A server that implements this extension MUST NOT accept the request
    to resume the session if the server_name extension contains a
    different name.  Instead, it proceeds with a full handshake to
    establish a new session.  When resuming a session, the server MUST
    NOT include a server_name extension in the server hello.

This seems to say that when the CloudFlare server receives the request with the same session ID but a different server_name, it's expected to create a new session, which it seems to not be doing. It seems to just be resuming the previous session and ignoring the new server_name provided.

I'm not sure if the last sentence implies the new session should not have the requested server_name associated with it, which would be just as bad, so I'm not willing to say NSS is doing the right thing, but CloudFlare definitely seems to be doing the wrong thing.

**Comment by vobe** [comment]
Found some sort of a fix by patching Pidgin 
```
--- a/libpurple/plugins/ssl/ssl-nss.c
+++ b/libpurple/plugins/ssl/ssl-nss.c
@@ -544,8 +544,10 @@ ssl_nss_connect(PurpleSslConnection *gsc)
        if (gsc->verifier != NULL)
                SSL_AuthCertificateHook(nss_data->in, ssl_auth_cert, NULL);
 
-       if(gsc->host)
+       if(gsc->host) {
                SSL_SetURL(nss_data->in, gsc->host);
+               SSL_SetSockPeerID(nss_data->in, gsc->host);
+       }
 
 #if 0
        /* This seems like it'd the be the correct way to implement the
```

It seems to work so far (but we should probably include the port too, as the doc says). Also, it might be worth reporting to CloudFlare before sending a bug report to pidgin .

**Comment by syldrathecat** [comment]
Turns out this is caused by a bug that was very recently fixed in NSS, but we need to include a workaround since the old version is likely to hang around for a while.

Created this proof of concept for an "nssfix" plugin, which when loaded along-side the ssl-nss plugin, will apply the same SetSockPeerID workaround as in the comment above: https://gist.github.com/syldrathecat/e0c8d78d43ebed092907

It could easily be integrated in to the current Makefile for both Windows and Linux, and shouldn't cause any future problems if left around in someone's plugin directory. Not sure if it's possible to make a plugin always load, but maybe the f-list plugin could attempt to load it or make it a dependency?

Unfortunately it brings in libnss as a build dependency for the project. Also if not made to load before other ssl plugins, or if ssl-nss is unloaded somehow, it will probably crash.

**Comment by syldrathecat** [comment]
There's been multiple people who have come to the channel solely to ask about this bug. The last one was a Linux user not building from source, so there was no testing build on hand to give them.

I think this is the single biggest thing affecting any user using multiple characters right now (Invalid JSON error when connecting).

It'd be nice if the nssfix branch could be merged in to master ASAP, if no problems have been found with it by now, so anyone keeping up with the snapshots will receive the fix. I think it'd be a really good time to start working on getting a new release out with the fix in it too (or potentially an 0.5.1 release with only the fix).
